### DISCUSSION TITLE
Discerning rules that aren't enabled and don't raise errors

### DISCUSSION THREAD

**Discussion by jamesbraza** [astral-sh/ruff#7796] [Q&A] [answered]
I have `select`/`disable` lists populated, and it's not a universal list of all rules.

Is there some simple way to discern rules I can add to the `select` list that either:
- Don't raise any errors
- After autofixing don't raise any errors

**Comment by charliermarsh** [comment]
There isn't a great way to do this right now (to my knowledge), but it was discussed once in https://github.com/astral-sh/ruff/issues/3560.

**Comment by charliermarsh** [comment]
I thought maybe you could do `ruff check --select ALL --statistics`, but that omits any rules with 0 violations...

**Reply by jamesbraza** [reply]
I like the way you're thinking here!! If `--statistics` showed rules with 0 violations, it would be very easy to `grep` from there.

To find autofixable ones, could just do `ruff check --select=ALL --fix`, then run `ruff check --select ALL --statistics` to get fully autofixable rules afterwards

**Reply by charliermarsh** [reply]
It feels wasteful to show all 0-violation rules all the time, but also feels like it doesn't warrant its own command-line option. We could consider including them only when `--format=json`, since that's already intended for machine consumption...

**Reply by jamesbraza** [reply]
That sounds like a good idea!  Would enable something like `ruff check --select ALL --fix && ruff check --select ALL --statistics --format=json | jq "." | grep "0"`

There's also probably some `sed` magic involving listing all rules, then subtracting rules from `ruff check --select ALL --statistics`

**Reply by jamesbraza** [reply]
Alternately, one can accomplish this with a Python script:

1. `ruff rule --format=json --all`: `set` of all ruff rules
2. `ruff check --select=ALL --fix && ruff check --select=ALL --statistics --format=json`: `set` of all erroring rules
3. Compute a set difference and print as a sorted list

**Reply by jamesbraza** [reply, marked as answer]
@charliermarsh I actually made this, shown in this issue: https://github.com/astral-sh/ruff/issues/7805

**Reply by jamesbraza** [reply]
Okay it ended up having a lot of little details to figure out, so I massively upgraded it lol
